### PR TITLE
Added sym_numel

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82209
* __->__ #82058
* #82054
* #82052

### CODE REVIEW

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: c10/core/TensorImpl.cpp [hunk 1]
```diff
@@ -669,7 +678,7 @@ void TensorImpl::Extend(int64_t num, float growthPct) {
           sizes_and_strides_.size_at_unchecked(0).as_int_unchecked() *
           (1 + growthPct / 100))));
   auto oldData = std::move(storage_.data_ptr());
-  auto oldSize = numel_;
+  auto oldSize = numel_.expect_int();
```
Comment (ezyang) [inline c10/core/TensorImpl.cpp:681]: this can be unchecked as has_symbolic_sizes_strides_ is already tested

File: c10/core/TensorImpl.cpp [hunk 2]
```diff
@@ -745,7 +754,7 @@ void TensorImpl::Reshape(const std::vector<int64_t>& dims) {
     new_size *= d;
   }
   TORCH_CHECK(
-      new_size == numel_,
+      new_size == numel_.expect_int(),
```
Comment (ezyang) [inline c10/core/TensorImpl.cpp:757]: this can be unchecked as has_symbolic_sizes_strides_ is already tested

File: c10/core/TensorImpl.h [hunk 3]
```diff
@@ -564,6 +564,22 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
 
   virtual c10::SymIntArrayRef sym_sizes_custom() const;
 
+  // TODO: make it non-virtual after a change to XLA
+  virtual c10::SymInt sym_numel() const {
```
Comment (ezyang) [inline c10/core/TensorImpl.h:568]: I don't see why this has to be virtual. XLA doesn't have this method. So why virtualize it?

File: c10/core/TensorImpl.h [hunk 4]
```diff
@@ -564,6 +564,22 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
...
+    if (C10_UNLIKELY(
+            sizes_strides_policy_ >=
+            static_cast<uint8_t>(SizesStridesPolicy::CustomSizes))) {
+      return sym_numel_custom();
+    }
+    return sym_numel_default();
+  }
+
+  inline c10::SymInt sym_numel_default() const {
+    return c10::SymInt(numel_);
```
Comment (ezyang) [inline c10/core/TensorImpl.h:578]: The conversion shouldn't be necessary, you already have a symint

File: c10/core/TensorImpl.h [hunk 5]
```diff
@@ -748,7 +764,7 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
 #ifdef DEBUG
     TORCH_INTERNAL_ASSERT(compute_numel() == numel_);
 #endif
-    return numel_;
+    return numel_.expect_int();
```
Comment (ezyang) [inline c10/core/TensorImpl.h:767]: This should be unchecked, if you have symbolic sizes you never call into the default impl

File: c10/core/TensorImpl.h [hunk 6]
```diff
@@ -1933,9 +1949,10 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
       // We can reuse the existing buffer if the current data does not have
       // a special destructor and the new data doesn't have a special
       // constructor.
-      if (numel_ == 0 ||
+      if (numel_.expect_int() == 0 ||
```
Comment (ezyang) [inline c10/core/TensorImpl.h:1952]: I hope this is getting CSE'd; otherwise you need t hoist this to the top

File: caffe2/core/tensor.h [hunk 7]
```diff
@@ -433,6 +433,10 @@ class TORCH_API Tensor final {
     return impl_->sym_sizes();
   }
 
+  inline c10::SymInt sym_numel() const {
+    return impl_->sym_numel();
+  }
```
Comment (ezyang) [inline caffe2/core/tensor.h:438]: There's no reason to add the method here (same with the others). Legacy Caffe2 interface

**Comment by Chillee** [comment]
Closed in favor of https://github.com/pytorch/pytorch/pull/82374

**Pull request closed without merging**